File: java/micronaut-postgres/app/src/main/java/com/example/service/NotificationClient.java
package com.example.service;

import com.fasterxml.jackson.databind.ObjectMapper;
import io.micronaut.context.annotation.Value;
import jakarta.inject.Singleton;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.Map;

@Singleton
public class NotificationClient {

    private static final Logger LOG = LoggerFactory.getLogger(NotificationClient.class);

    private final HttpClient httpClient = HttpClient.newHttpClient();
    private final ObjectMapper objectMapper = new ObjectMapper();
    private final String notifyUrl;

    public NotificationClient(@Value("${notify.url:`http://localhost:8081`}") String notifyUrl) {
        this.notifyUrl = notifyUrl;
    }

    public void notify(Map<String, Object> payload) {
        try {
            String json = objectMapper.writeValueAsString(payload);
            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create(notifyUrl + "/notify"))
                    .header("Content-Type", "application/json")
                    .POST(HttpRequest.BodyPublishers.ofString(json))
                    .build();
            HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
            LOG.debug("Notify response: {}", response.statusCode());
        } catch (Exception e) {
            LOG.warn("Failed to notify: {}", e.getMessage());
        }
    }
}

File: kotlin/ktor-postgres/app/src/main/kotlin/com/example/service/NotificationClient.kt
package com.example.service

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.encodeToString
import org.slf4j.LoggerFactory
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse

class NotificationClient(private val notifyUrl: String) {

    private val logger = LoggerFactory.getLogger(NotificationClient::class.java)
    private val httpClient = HttpClient.newHttpClient()

    suspend fun notify(payload: Map<String, String>) {
        withContext(Dispatchers.IO) {
            val json = Json.encodeToString(payload)
            val request = HttpRequest.newBuilder()
                .uri(URI.create("$notifyUrl/notify"))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(json))
                .build()
            val response = httpClient.send(request, HttpResponse.BodyHandlers.ofString())
            logger.info("Notification sent: status={}", response.statusCode())
        }
    }
}
